implement a function to convert a storage to copy-on-write

dagitses opened the pull request:
implement a function to convert a storage to copy-on-write

Summary:
This will be used in the _lazy_clone() operator as well as reshape().

Test Plan: 100% coverage of reachable lines.

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/pytorch/pytorch/pull/100819).
* #100821
* #100820
* __->__ #100819

cc @ezyang @bhosmer @smessmer @ljk53 @bdhirsh

ezyang commented on c10/core/impl/cow/try_ensure.cpp:
```diff
@@ -0,0 +1,102 @@
+auto make_data_ptr(at::DataPtr const& data_ptr, cow::Context& ctx)
+    -> at::DataPtr {
+  return at::DataPtr(
+      data_ptr.get(), &ctx, cow::delete_context, data_ptr.device());
```
I was expecting to see some reference to the original data pointers deleter here?

ezyang replied:
Looks like it's handled below

ezyang commented on c10/core/impl/cow/try_ensure.cpp:
```diff
@@ -0,0 +1,102 @@
+    TORCH_INTERNAL_ASSERT(original_ctx.get() == data_ptr.get());
+
+    std::unique_ptr<cow::Context, AssertFalseDeleter> ctx(
+        new cow::Context(std::move(original_ctx)));
```
Why bother putting this in a unique_ptr :think:

ezyang replied:
But it's not really exception safe. Your deleter is AssertFalseDeleter, so if an exception is thrown, you'll assert false and terminate lol

dagitses replied:
True, but in some ways an assertion is better than a leak. Either way, it's reordered now and I think it's mostly exception safe, addresses the deleter and the cast_context as well.

ezyang commented on c10/core/impl/cow/try_ensure.cpp:
```diff
@@ -0,0 +1,102 @@
+    storage.set_data_ptr_noswap(make_data_ptr(data_ptr, *ctx));
+
+    // Save this for the result.
+    ctx->increment_refcount();
```
NB: refcount is 2 now!

dagitses committed:
implement a function to convert a storage to copy-on-write

Summary:
This will be used in the _lazy_clone() operator as well as reshape().

Test Plan: 100% coverage of reachable lines.

dagitses commented:
@pytorchbot merge

jeanschmidt commented:
@pytorchbot revert -m "added tests are breaking internal builds" -c nosignal

dagitses commented:
> @pytorchbot revert -m "added tests are breaking internal builds" -c nosignal

Hey Jean, sorry for the trouble. What was the nature of the failure?

dagitses commented:
> > @pytorchbot revert -m "added tests are breaking internal builds" -c nosignal
> 
> Hey Jean, sorry for the trouble. What was the nature of the failure?

@jeanschmidt

dagitses commented:
@ezyang Do you think I should just try landing this again without unit tests :P?

ezyang commented:
Now that you're on the other side, I think you should try not to add new test files haha
